format-xls/src/main/java/io/cdap/plugin/format/xls/input/XlsRowConverter.java

```diff
@@ -0,0 +1,167 @@
+
+  XlsRowConverter(FormulaEvaluator evaluator) {
+    this.evaluator = evaluator;
+  }
+
+  /**
+   * Converts a row of XLS cells to a StructuredRecord.
+   * Returns null if the row is null or empty.
+   */
+  public StructuredRecord.Builder convert(Row row, Schema outputSchema) {
+    if (row == null) {
+      return null;
+    }
+    boolean isRowEmpty = true;
+    StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+    List<Schema.Field> fields = outputSchema.getFields();
+    for (int cellIndex = 0; cellIndex < row.getLastCellNum(); cellIndex++) {
+      if (cellIndex >= fields.size()) {
+        throw new IllegalArgumentException(
+                String.format("Schema contains fewer fields than the number of columns in the excel file. " +
+                        "Schema fields: %s, Excel columns: %s", fields.size(), row.getLastCellNum()));
+      }
+      Cell cell = row.getCell(cellIndex, Row.MissingCellPolicy.RETURN_BLANK_AS_NULL);
+      if (cell == null) {
+        // Blank cells are skipped, builder will set null for the field, no processing needed.
+        continue;
+      }
+      Schema.Field field = fields.get(cellIndex);
+      Schema.Type type = field.getSchema().isNullable() ?
+              field.getSchema().getNonNullable().getType() : field.getSchema().getType();
+      Object cellValue;
+      switch (type) {
+        case STRING:
+          cellValue = getCellAsString(cell);
+          break;
+        case DOUBLE:
+          cellValue = getCellAsDouble(cell);
+          break;
+        case BOOLEAN:
+          cellValue = getCellAsBoolean(cell);
+          break;
+        default:
+          throw new IllegalArgumentException(
+                  String.format("Field '%s' is of unsupported type '%s'. Supported types are: %s",
+                          field.getName(), type, "string, double, boolean"));
+      }
+      if (cellValue == null) {
+        continue;
+      }
+      builder.set(field.getName(), cellValue);
+      isRowEmpty = false;
+    }
+    if (isRowEmpty) {
+      return null;
+    }
+    return builder;
+  }
+
+  private CellType getCellType(Cell cell) {
+    CellType cellType = cell.getCellType();
+    if (cellType == CellType.FORMULA) {
+      try {
+        cellType = cell.getCachedFormulaResultType();
+      } catch (Exception e) {
+        cellType = evaluator.evaluateFormulaCell(cell);
+      }
+    }
+    return cellType;
+  }
+
+  private String getCellAsString(Cell cell) {
+    CellType cellType = getCellType(cell);
+
+    switch (cellType) {
+      case NUMERIC:
+        if (DateUtil.isCellDateFormatted(cell)) {
+          return dataFormatter.formatCellValue(cell);
+        }
+        return Double.toString(cell.getNumericCellValue());
+      case STRING:
+        return cell.getRichStringCellValue().getString();
+      case BOOLEAN:
+        return cell.getBooleanCellValue() ? "TRUE" : "FALSE";
+      case BLANK:
+      case ERROR:
```

Comment:
are there any other CellType values (like date?)

Reply:
No , dates are stored as Numeric, there is a util function to check if the numeric cell is formatted as date.